### PR TITLE
[Feature] url-input

### PR DESCRIPTION
**Pull request by ITZVGcGPmO**
**Added**
- `/skin <url>`
- more lang entries

**TODO**
- permission nodes :thinking:
- skin blacklisting? :thinking:
- redo skin storage and thus the entire plugin :thinking: (or a "whitespace prefix" workaround)

**needs verification**
- bStats seeing MS-API calls
- relog persistence (100Yr)

etc
- tacos 🌮

### CODE REVIEW

**Comment by ITZVGcGPmO** [comment]
You can file [issues](https://github.com/ITZVGcGPmO/SkinsRestorerX/issues) and [pulls](https://github.com/ITZVGcGPmO/SkinsRestorerX/pulls) at [ITZVGcGPmO/SkinsRestorerX](https://github.com/ITZVGcGPmO/SkinsRestorerX)

**Comment by riflowth** [comment]
Nice feature! Very suitable for SkinsRestorer. 👍

File: src/main/java/skinsrestorer/shared/utils/MetricsCounter.java [hunk 1]
```diff
@@ -8,65 +8,78 @@
...
     public static void setBackup_calls(int backup_calls) {
         backup_calls = backup_calls;
     }
-
     public static void incrBackup_calls() {
         backup_calls = backup_calls + 1;
     }
 
     public static void incrAPI(String url) {
+        System.out.println("[SkinsRestorer] API call: "+url);
```
Comment (McLive) [inline src/main/java/skinsrestorer/shared/utils/MetricsCounter.java]:
```suggestion
```

File: src/main/java/skinsrestorer/shared/utils/MineSkinAPI.java [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+                }
+            }
+            return "alex";
+        }
+        catch (Exception e) {}
+        return "steve";
+    }
+
+    public static Object genSkin(String url) throws MojangAPI.SkinRequestException {
+        return genSkin(url, null); }
```
Comment (McLive) [inline src/main/java/skinsrestorer/shared/utils/MineSkinAPI.java]:
```suggestion
        return genSkin(url, null);
    }
```

File: src/main/java/skinsrestorer/shared/utils/MineSkinAPI.java [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+                    int  green = (clr & 0x0000ff00) >> 8;
+                    int  blue  =  clr & 0x000000ff;
+                    if (alpha != 0 && (!(red == 255 || red == 0) || !(green == 255 || green == 0) || !(blue == 255 || blue == 0))) {
+                        return "steve";
+                    }
+                }
+            }
+            return "alex";
+        }
+        catch (Exception e) {}
```
Comment (McLive) [inline src/main/java/skinsrestorer/shared/utils/MineSkinAPI.java]:
```suggestion
        catch (Exception ignored) {
        }
```

File: src/main/java/skinsrestorer/shared/utils/MineSkinAPI.java [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+                output.close();
+                String outstr = "";
+                InputStream _is;
+                try {
+                    _is = con.getInputStream();
+                } catch (Exception e) {
+                    _is = con.getErrorStream();
+                }
+                DataInputStream input = new DataInputStream(_is);
+                for( int c = input.read(); c != -1; c = input.read() ) { outstr += (char)c; }
```
Comment (McLive) [inline src/main/java/skinsrestorer/shared/utils/MineSkinAPI.java]:
```suggestion
                for (int c = input.read(); c != -1; c = input.read()) {
                    outstr += (char)c;
                }
```

File: src/main/java/skinsrestorer/shared/utils/MineSkinAPI.java [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+                input.close();
+                return outstr;
+            } catch (Exception e) {}
```
Comment (McLive) [inline src/main/java/skinsrestorer/shared/utils/MineSkinAPI.java]:
```suggestion
            } catch (Exception ignored) {
            }
```

**Comment by ITZVGcGPmO** [comment]
Should be good to go! :smiley:

File: src/main/java/skinsrestorer/bukkit/commands/SkinCommand.java [hunk 6]
```diff
@@ -158,18 +156,40 @@ private boolean setSkin(CommandSender sender, Player p, String skin, boolean sav
...
+
+                // set custom skin name back to old one if there is an exception
+                if (save)
+                    SkinStorage.setPlayerSkin(p.getName(), oldSkinName != null ? oldSkinName : p.getName());
+            }
+            return false;
+        }
+        if (C.validUrl(skin)) {
+            try {
+                p.sendMessage(Locale.MS_UPDATING_SKIN);
```
Comment (McLive) [inline src/main/java/skinsrestorer/bukkit/commands/SkinCommand.java]:
```suggestion
                sender.sendMessage(Locale.MS_UPDATING_SKIN);
```

File: src/main/java/skinsrestorer/bungee/commands/SkinCommand.java [hunk 7]
```diff
@@ -158,26 +156,45 @@ private boolean setSkin(CommandSender sender, ProxiedPlayer p, String skin, bool
...
+                this.rollback(p, oldSkinName, save); // set custom skin name back to old one if there is an exception
+            } catch (Exception e) {
+                e.printStackTrace();
+                this.rollback(p, oldSkinName, save); // set custom skin name back to old one if there is an exception
+            }
+            return false;
+        }
+        if (C.validUrl(skin)) {
+            try {
+                p.sendMessage(new TextComponent(Locale.MS_UPDATING_SKIN));
```
Comment (McLive) [inline src/main/java/skinsrestorer/bungee/commands/SkinCommand.java]:
```suggestion
                sender.sendMessage(new TextComponent(Locale.MS_UPDATING_SKIN));
```

File: src/main/java/skinsrestorer/shared/utils/MetricsCounter.java [hunk 8]
```diff
@@ -8,65 +8,77 @@
...
         backup_calls = backup_calls;
     }
-
     public static void incrBackup_calls() {
         backup_calls = backup_calls + 1;
     }
 
     public static void incrAPI(String url) {
+
+        if (url.startsWith("https://api.mineskin.org/"))
```
Comment (McLive) [inline src/main/java/skinsrestorer/shared/utils/MetricsCounter.java]: Maybe we want to count https://api.minetools.eu/mineskin/ to mineskin requests?!

Comment (ITZVGcGPmO) [reply src/main/java/skinsrestorer/shared/utils/MetricsCounter.java]: i think this should be done server-side if possible, as minetools will cache any responses.

plus, minetools requests are already logged.

File: src/main/java/skinsrestorer/shared/storage/Locale.java [hunk 9]
```diff
@@ -21,17 +21,20 @@
...
     public static String SKIN_CLEAR_SUCCESS = "&e[&2SkinsRestorer&e] &2Your skin has been cleared.";
     public static String SKIN_CLEAR_ISSUER = "&e[&2SkinsRestorer&e] &2Skin cleared for player %player.";
 
+    public static String MS_UPDATING_SKIN = "&e[&2SkinsRestorer&e] &2Uploading skin, please wait!";
     public static String SUCCESS_UPDATING_SKIN = "&e[&2SkinsRestorer&e] &2Your skin has been updated.";
     public static String SUCCESS_UPDATING_SKIN_OTHER = "&e[&2SkinsRestorer&e] &2Skin updated for player %player.";
-    
+
     public static String ERROR_UPDATING_SKIN = "&e[&2SkinsRestorer&e] &4Error&8: &cAn error happened while updating your skin. Please try again later!";
+    public static String ERROR_MS_FULL = "&e[&2SkinsRestorer&e] &4MS Error&8: &cCouldn't upload skin. (Consider https://mineskin.org/account )!";
```
[the remark this answers is not in the document]
Comment (ITZVGcGPmO) [reply src/main/java/skinsrestorer/shared/storage/Locale.java]: would
```MS Error: No uploader accounts were available to fulfill your request. Try again later or consider https://mineskin.org/account```
be better?

I really want to get account signup promoted, when mineskin is lacking accounts. If that means some kiddos get confused, so be it. I think the risk is worth the reward. I would like to hear your opinion on this, and if we should do our part to promote mineskin account contribution to the end-users.

Comment (McLive) [inline src/main/java/skinsrestorer/shared/storage/Locale.java]: Looks like you removed your better message :o

File: src/main/java-templates/skinsrestorer/sponge/commands/SkinCommand.java [hunk 10]
```diff
@@ -162,18 +162,40 @@ private boolean setSkin(CommandSource source, Player p, String skin, boolean sav
...
+                source.sendMessage(plugin.parseMessage(e.getReason()));
+                // set custom skin name back to old one if there is an exception
+                if (save)
+                    SkinStorage.setPlayerSkin(p.getName(), oldSkinName != null ? oldSkinName : p.getName());
+            }
+            return false;
+        }
+        if (C.validUrl(skin)) {
+            try {
+                p.sendMessage(plugin.parseMessage(Locale.MS_UPDATING_SKIN));
```
Comment (McLive) [inline src/main/java-templates/skinsrestorer/sponge/commands/SkinCommand.java]:
```suggestion
                source.sendMessage(plugin.parseMessage(Locale.MS_UPDATING_SKIN));
```

File: src/main/java-templates/skinsrestorer/velocity/command/SkinCommand.java [hunk 11]
```diff
@@ -168,26 +168,51 @@ private boolean setSkin(CommandSource source, Player p, String skin, boolean sav
...
-            } catch (Exception e) {
-                e.printStackTrace();
-
-                // set custom skin name back to old one if there is an exception
-                this.rollback(p, oldSkinName, save);
+                return false;
             }
+            if (C.validUrl(skin)) {
+                try {
+                    p.sendMessage(plugin.deserialize(Locale.MS_UPDATING_SKIN));
```
Comment (McLive) [inline src/main/java-templates/skinsrestorer/velocity/command/SkinCommand.java]:
```suggestion
                    source.sendMessage(plugin.deserialize(Locale.MS_UPDATING_SKIN));
```

**Pull request merged**